contracts/starknet/teleport_join.cairo

```diff
@@ -0,0 +1,938 @@
+    #         }
+    let (_teleport : TeleportStatus) = _teleports.read(hashGUID)
+    let pending : Uint256 = _teleport.pending
+    let (pending_null) = eq_0(pending)
+    let (over_ceiling) = le_int(line_, debt_)
+    let (nothing_to_withdraw) = either(over_ceiling, pending_null)
+
+    if nothing_to_withdraw == 1:
+        Mint.emit(hashGUID, teleportGUID, Uint256(0, 0), max_fee_percentage, operator_fee, caller)
+        return (Uint256(0, 0), Uint256(0, 0))
+    end
+
+    # uint256 amtToTake = _min(
+    #                                 pending,
+    #                                 uint256(int256(line_) - debt_)
+    #                             );
+    let (under_ceiling : Uint256) = _sub(line_, debt_)
+    let (amt_to_take : Uint256) = _min(pending, under_ceiling)
+
+    # uint256 fee = vatLive ? FeesLike(fees[teleportGUID.sourceDomain]).getFee(teleportGUID, line_, debt_, pending, amtToTake) : 0;
+    #         require(fee <= maxFeePercentage * amtToTake / WAD, "TeleportJoin/max-fee-exceed");
+
+    let (domain_fees) = _fees.read(teleportGUID.source_domain)
+    let (teleport_fee) = FeesLike.getFee(
+        domain_fees, teleportGUID, line_, debt_, pending, amt_to_take
+    )
+    let (fee : Uint256) = mul(Uint256(vat_live, 0), teleport_fee)
+
+    # // No need of overflow check here as amtToTake is bounded by teleports[hashGUID].pending
+    #         // which is already a uint248. Also int256 >> uint248. Then both castings are safe.
+    #         debt[teleportGUID.sourceDomain] +=  int256(amtToTake);
+    #         teleports[hashGUID].pending     -= uint248(amtToTake);
+    let (add_debt : Uint256) = add(debt_, amt_to_take)
+    _debts.write(teleportGUID.source_domain, add_debt)
+    let (pending_update : Uint256) = sub(_teleport.pending, amt_to_take)
+
+    let new_status = TeleportStatus(blessed=_teleport.blessed, pending=pending_update)
+    _teleports.write(hashGUID, new_status)
+
+    # if (debt_ >= 0 || uint256(-debt_) < amtToTake) {
+    #             uint256 amtToGenerate = debt_ < 0
+    #                                     ? uint256(int256(amtToTake) + debt_) // amtToTake - |debt_|
+    #                                     : amtToTake;
+    #             // amtToGenerate doesn't need overflow check as it is bounded by amtToTake
+    #             vat.slip(ilk, address(this), int256(amtToGenerate));
+    #             vat.frob(ilk, address(this), address(this), address(this), int256(amtToGenerate), int256(amtToGenerate));
+    #             // Query the actual value as someone might have repaid debt without going through settle (if vat.live == 0 prev frob will revert)
+    #             (, art) = vat.urns(ilk, address(this));
+    #         }
+
+    let (self) = get_contract_address()
+
+    let (debt_pos) = _ge_0(debt_)
+    let (minus_debt) = uint256_neg(debt_)
+    let (mint_needed) = le_int(minus_debt, amt_to_take)
+    let (should_mint) = either(debt_pos, mint_needed)
+    if should_mint == 1:
```

Comment:
Not sure if it is worth the complexity, `should_mint` can be optimized via lazy evaluation.

Reply:
how would you go about it in this case ? will try smth

Reply:
I would leave a TODO. Not worth the complexity.